### ISSUE TITLE
`cursor id XXX not found` when streaming data from secondary

### ISSUE THREAD

**Issue by ornicar** [ReactiveMongo/ReactiveMongo#1185]
- "org.reactivemongo" %% "reactivemongo" % "1.1.0-275c4ca-RC7-SNAPSHOT",
- "org.reactivemongo" %% "reactivemongo-akkastream" % "1.1.0-RC6",
- MongoDB 5.0.13 (and 4.4.13)
- Linux 5.19.11-arch1-1
- OpenJDK 17.0.5
- scala 3.2.

Streaming data from a secondary results in the stream closing with an error after some documents.

```scala
coll
      .find(BSONDocument(), Some(BSONDocument("_id" -> true)))
      .cursor[BSONDocument](ReadPreference.secondary)
      .documentSource(nbDocs)
      .mapConcat(d => d.int("_id").toList)
      .throttle(50, 1.second)
      .runWith(Sink.foreach(println))
```

```
Exception in thread "main" DatabaseException['cursor id 8910965325409551864 not found' (code = 43)]
```

I made a [minimal reproduction program](https://github.com/ornicar/rm-issue) with a README.

To reproduce you need a replicaset with at least one secondary.

**Comment by ornicar** [comment]
We're facing another problem, where streaming documents from the primary results in the document stream hanging (not terminating with an error) after some variable amount of documents.
So far this can only be seen within lila, not in a simple program. I'm working on reproducing it outside lila.

**Comment by tardigradeus** [comment]
Hello everyone,

I am facing the same problem and investigated. I thing i can provide more information:

- The bug was introduced in 1.1.0-RC6 from 2c1e6e24.
- The bug occurs when the number of documents exceeds the default batch size (101).

I think with a replicaset and a `writeConcern=secondaryPreferred` the Cursor jumps from one node to another node while fetching the next batch.

This was introduced with the `RequestMaker` refactor that erased the `channelIdHint`.

#### Before, the [code](https://github.com/ReactiveMongo/ReactiveMongo/commit/2c1e6e24db8849abccad71832fc219b052d82d3f#diff-bff81c0da72a381090bb898e75023209658324bea890f767ff0d7f7bcbd1cbe4L331) was building a `requestMaker` and gives the previous channelIdHint: 
```
val (op, cmd) = getMoreOpCmd(reply.cursorID, ntr)

logger.trace(s"Asking for the next batch of $ntr documents on cursor #${reply.cursorID}, after ${nextOffset}: $op")

def req = new ExpectingResponse(
  requestMaker = RequestMaker(CommandKind.Query, op, cmd,
    readPreference = preference,
    channelIdHint = Some(response.info.channelId)),
  pinnedNode = transaction.flatMap(_.pinnedNode))
```

#### After the commit the reqMaker [comes](https://github.com/ReactiveMongo/ReactiveMongo/commit/2c1e6e24db8849abccad71832fc219b052d82d3f#diff-bff81c0da72a381090bb898e75023209658324bea890f767ff0d7f7bcbd1cbe4R478-R483) from the `getMoreOpCmd` function: 
```
val reqMaker = getMoreOpCmd(reply.cursorID, ntr)

logger.trace(s"Asking for the next batch of $ntr documents on cursor #${reply.cursorID}, after ${nextOffset}: ${reqMaker.op}")

def req = new ExpectingResponse(
  requestMaker = reqMaker,
  pinnedNode = transaction.flatMap(_.pinnedNode))
``` 

In the [body](https://github.com/ReactiveMongo/ReactiveMongo/commit/2c1e6e24db8849abccad71832fc219b052d82d3f#diff-bff81c0da72a381090bb898e75023209658324bea890f767ff0d7f7bcbd1cbe4R238) of the `getMoreOpCmd` function the `RequestMaker` returned has the `channelIdHint` value fixed to `None`

```
RequestMaker(
  kind = CommandKind.Query,
  op = moreQry,
  section = buffer.buffer,
  readPreference,
  channelIdHint = None,
  callerSTE = Seq.empty
)
``` 

I will submit a PR to propose a fix.

Do you agree with my investigation ? @cchantep

**Comment by martin-lettner** [comment]
Hello @tardigradeus!

Thanks a lot for your thorough analysis and the pull request with the proposed fix!

Unfortunately we experience the same issue _(when reading with a `cursor` from a replicaset with the read preference _secondary preferred_ the call `collect` returns just up to 101 (default batch size) documents._) with the versions **1.1.0-RC7** to **1.1.0-RC12**. (_With the version **1.1.0-RC6** we had some dependency issues with the logging framework._)

@cchantep Is there already a planned date for the next release candidate which will contain the merged pull request #1299?

**Comment by cchantep** [comment]
Will have a look. Everything was blocked due to Sonatype changing its repository management.
